packages/pug-lexer/index.js

```diff
@@ -990,172 +1046,244 @@ Lexer.prototype = {
+      }
+    }
+    
+    line = this.lineno;
+    col = this.colno;
+    
+    // start looping through the value
+    for (; i < str.length; i++) {
+      // if the character is in a string or in parentheses/brackets/braces
+      if (!(state.isNesting() || state.isString())){
+        
+        if (this.whitespaceRe.test(str[i])) {
+          done = false;
+          
+          // find the first non-whitespace character
+          for (x = i; x < str.length; x++) {
+            if (!this.whitespaceRe.test(str[x])) {
+              // if it is a JavaScript punctuator, then assume that it is
+              // a part of the value
+              if((!characterParser.isPunctuator(str[x]) || quoteRe.test(str[x]) || str[x] === ':') && this.assertExpression(val, true)){
+                done = true;
               }
+              break;
             }
           }
-          // if there's no whitespace and the character is not ',', the
-          // attribute did not end.
-          return str[i] === ',';
-        } else if (loc === 'value') {
-          // if the character is in a string or in parentheses/brackets/braces
-          if (state.isNesting() || state.isString()) return false;
-
-          // if the current value expression is not valid JavaScript, then
-          // assume that the user did not end the value.  To enforce this,
-          // we call `self.assertExpression(val, true)`, but since the other
-          // tests are much faster, we run the other tests first.
-
-          if (whitespaceRe.test(str[i])) {
-            // find the first non-whitespace character
-            for (var x = i; x < str.length; x++) {
-              if (!whitespaceRe.test(str[x])) {
-                // if it is a JavaScript punctuator, then assume that it is
-                // a part of the value
-                return (!characterParser.isPunctuator(str[x]) || quoteRe.test(str[x]) || str[x] === ':') && self.assertExpression(val, true);
-              }
-            }
+          
+          // if everything else is whitespace, return now so last attribute
+          // does not include trailing whitespace
+          if(done || x === str.length){
+            break;
           }
-          // if there's no whitespace and the character is not ',', the
-          // attribute did not end.
-          return str[i] === ',' && self.assertExpression(val, true);
+        }
+        
+        // if there's no whitespace and the character is not ',', the
+        // attribute did not end.
+        if(str[i] === ',' && this.assertExpression(val, true)){
+          break;
         }
       }
+      
+      state = characterParser.parseChar(str[i], state);
+      val += str[i];
+      
+      if (str[i] === '\n') {
+        line++;
+        col = 1;
+      } else {
+        col++;
+      }
+    }
+    
+    this.assertExpression(val);
+    
+    this.lineno = line;
+    this.colno = col;
+    
+    tok.val = val;
+    tok.mustEscape = escapeAttr;
```

Comment:
Rather than taking and mutating a token. I feel if would be clearer if this function returned `{val, mustEscape, remainingSource}`